contrib/babelfishpg_common/src/datetime.c

```diff
@@ -46,6 +47,456 @@ PG_FUNCTION_INFO_V1(datetime_mi_datetime);
+	 * beginning of the DATE field.
+	 * Also, if it is in ISO 8601 format, we need to check whether
+	 * input str is matching 'YYYY-MM-DDThh:mm:ss[.mmm]' format.
+	 */
+	for (i = start_idx; i < last_idx; i++)
+	{
+		if (ftype[i] == DTK_STRING)
+		{
+			int j = i-1, temp_int;
+			char *temp;
+
+			ftype[i] = DecodeSpecial(i, field[i], &temp_int);
+
+			if (ftype[i] == ISOTIME && (contains_extra_spaces || num_colons < 2))
+				return 1;
+			
+			else if (ftype[i] == ISOTIME)
+			{
+				char *regex_exp = "^[0-9]{4}[-][0-9]{2}[-][0-9]{2}[T][0-9]{2}[:][0-9]{2}[:][0-9]{2}([.][0-9]{1,3})?$";
+				
+				if (!match_regex(orig_str, regex_exp))
+					return 1;
+			}
+
+			if(ftype[i] != MONTH)
+			{
+				ftype[i] = DTK_STRING;
+				continue;
+			}
+
+			contains_text_month = true;
+
+			if(!check_regex_for_text_month(str, DATE_TIME))
+				return 1;
+
+			tm->tm_mon = temp_int;
+
+			while (j>=0)
+			{
+				/* Swap ftype entries */
+				temp_int = ftype[j];
+				ftype[j] = ftype[j+1];
+				ftype[j+1] = temp_int;
+
+				/* Swap field entries */
+				temp = field[j];
+				field[j] = field[j+1];
+				field[j+1] = temp;
+
+				j--;
+			}
+		}
+		else if (ftype[i] == DTK_NUMBER && !date_exists)
+		{
+			int field_len = strlen(field[i]);
+			if (last_idx - start_idx > 2)
+				continue;
+
+			if (field_len == 4)
+			{
+				tm->tm_year = atoi(field[i]);
+				*is_year_set = true;
+			}
+			else if (field_len == 2 && time_idx == 2)
+			{
+				int temp = atoi(field[i]);
+				tm->tm_year = (temp < 50) ? (2000 + temp) : (1900 + temp);
+				*is_year_set = true;
+			}
+			else if (last_idx - start_idx != 1 || (field_len != 8 && field_len != 6))
+				return 1;
+
+
+		}
+		else if (ftype[i] == DTK_DATE)
+		{
+			/*
+			 * Check whether the field contains any alphabet.
+			 * If yes, it might contain text month. 
+			 * If no alphabet are found, we need to check whether the
+			 * length of the DATE field is less than 10. Throw error if it
+			 * exceeds.
+			 */
+			char *cp = field[i];
+			/* regex expression for "[m]m-yyyy-[d]d" format */
+			char *regex_expr = "^[0-9]{1,2}[-|.|/][0-9]{4}[-|.|/][0-9]{1,2}$";
+
+			while (cp != NULL && *cp != '\0' && !isalpha(*cp))
+				cp++;
+
+			if (cp != NULL && isalpha(*cp))
+				continue;
+			
+			/*
+			 * If the DATE is in mm-yyyy-dd format, DecodeDateTime
+			 * throws error, hence we need to rearrange it the DATE string
+			 * to mm-dd-yyyy format
+			 */
+			
+			if (match_regex(field[i], regex_expr))
+			{
+				int month = 0, year = 0, day = 0;
+				cp = field[i];
+
+				month = strtoi64(cp, &cp, 10);
+				cp++;
+				year = strtoi64(cp, &cp, 10);
+				cp++;
+				day = strtoi64(cp, &cp, 10);
+
+				sprintf(field[i], "%d.%d.%d", month, day, year);
+			}
+
+			if (strlen(field[i]) > 10)
+				return 1;
+		}
+	}
+
+	if (!am_pm || contains_time)
+		return (contains_text_month == true) ? (number_fields > 2) : (number_fields > 1);
+	else
+		return (contains_text_month == true) ? (number_fields > 3) : (number_fields > 1);
```

Comment:
1. Can simply do  `(contains_text_month) ?` 
2. It's not entirely clear what we are doing here, probably a comment would help.